src/dataplane-protocol/src/smartstream.rs

```diff
@@ -0,0 +1,41 @@
+pub use encoding::SmartStreamUserError;
+pub use encoding::SmartStreamOutput;
+
+pub type Result<T> = std::result::Result<T, SmartStreamUserError>;
+
+mod encoding {
+    use crate::record::Record;
+    use fluvio_protocol::{Encoder, Decoder};
+    use std::fmt;
+
+    /// A type used to capture and serialize errors from within a SmartStream
+    #[derive(Debug, Default, Clone, PartialEq, Encoder, Decoder)]
+    pub struct SmartStreamUserError {
+        rendered: String,
```

Comment:
we should at least model error similar to: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/RuntimeError

Reply:
So I sort of tried to do something like that where all of the info is captured in a structural way, but unfortunately there is a good deal of magic with the `eyre` crate that actually captures this information, and it does not have an interface that allows you to extract that info in a structured way, only as Display or Debug text.

I was thinking that in the future we could dig down and just learn the primitives that eyre uses to e.g. capture the location of an error, but that is a bit out of scope for the time being. My imagined upgrade path for this would be:

- Add a new variant to the top-level SmartStreamError, `SmartStreamError::UserErrorV2(...)`
- Write `SmartStreamUserErrorV2` that can capture structured errors using the same primitives `eyre` uses